controllers/flinkcluster_submit_job_script.go

```diff
@@ -22,87 +22,134 @@ package controllers
+# When succeeded:
+#
+# jobID: ec74209eb4e3db8ae72db00bd7a830aa
+# message: |
+#   Successfully submitted!
+#   /opt/flink/bin/flink run --jobmanager flinkjobcluster-sample-jobmanager:8081 --class org.apache.flink.streaming.examples.wordcount.WordCount --parallelism 2 --detached ./examples/streaming/WordCount.jar --input ./README.txt
+#	Starting execution of program
+#	Printing result to stdout. Use --output to specify output path.
+#	Job has been submitted with JobID ec74209eb4e3db8ae72db00bd7a830aa
+#
+# When failed (no jobID):
+#
+# message: |
+#   Aborted submit because JobManager is unavailable.
+#   Checking job manager to be ready. Will check success of 2 API calls for stable job submission.
+#	curl -sS "http://t3-jobmanager:8081/jobs"
+#	curl: (7) Failed to connect to t3-jobmanager port 8081: Connection refused
+#   ...
 
 set -euo pipefail
 
-JOB_MANAGER="$2"
-
-function list_jobs() {
-	for i in {1..10}; do
-		if /opt/flink/bin/flink list -a --jobmanager "${JOB_MANAGER}" 2>&1; then
-			return 0
-		else
-			sleep 5
-		fi
-	done
-
-	echo "Failed to list jobs." >&2
-	return 1
+readonly TERM_LOG="/dev/termination-log"
+
+function check_jm_ready() {
+    # Waiting for 5 mins.
+    local -r MAX_RETRY=60
+    local -r RETRY_INTERVAL=5s
+    local -r REQUIRED_SUCCESS_NUMBER=2
+    local success_count=0
+
+    echo_log "Checking job manager to be ready. Will check success of ${REQUIRED_SUCCESS_NUMBER} API calls for stable job submission." "job_check_log"
+    for ((i = 1; i <= MAX_RETRY; i++)); do
+        echo_log "curl -sS \"http://${FLINK_JM_ADDR}/jobs\"" "job_check_log"
+        if curl -sS "http://${FLINK_JM_ADDR}/jobs" 2>&1 | tee -a job_check_log; then
+            ((success_count++))
+            echo_log "\nSuccess ${success_count}/${REQUIRED_SUCCESS_NUMBER}" "job_check_log"
+            if ((success_count < REQUIRED_SUCCESS_NUMBER)); then
+                echo_log "\nWaiting..." "job_check_log"
+                sleep "${RETRY_INTERVAL}"
+                continue
+            fi
+            echo_log "\nJob manager is ready now. Tried ${i} time(s), every ${RETRY_INTERVAL} and succeeded ${success_count} time(s)." "job_check_log"
+            return 0
+        else
+            echo_log "\nWaiting..." "job_check_log"
+        fi
+        sleep "${RETRY_INTERVAL}"
+    done
+
+    echo_log "\nReached max retry count(${MAX_RETRY}) to check job manager status." "job_check_log"
+    echo_log "Aborted to submit job." "job_check_log"
+
+    format_log_message "Aborted submit because JobManager is unavailable." "job_check_log" >>"${TERM_LOG}"
+
+    return 1
 }
 
-function check_existing_jobs() {
-	echo "Checking existing jobs..."
-	list_jobs
-	if list_jobs | grep -e "(SCHEDULED)" -e "(CREATED)" -e "(SUSPENDED)" -e "(FINISHED)" -e "(FAILED)" -e "(CANCELED)" \
-		-e "(RUNNING)" -e "(RESTARTING)" -e "(CANCELLING)" -e "(FAILING)" -e "(RECONCILING)"; then
-		echo "Found an existing job, skip resubmitting..."
-		return 0
-	fi
-	return 1
```

Comment:
@elanv, could you please explain why you removed this functionality?
We use HA mode. So that, when the cluster is recreated (removed and and deployed again), the newly created jobmanager is able to restore cluster's state and the job. And in the same time the job submitter submitted the same job again. We ended up with two jobs.

Reply:
> @elanv, could you please explain why you removed this functionality?
> We use HA mode. So that, when the cluster is recreated (removed and and deployed again), the newly created jobmanager is able to restore cluster's state and the job. And in the same time the job submitter submitted the same job again. We ended up with two jobs.

The job is now being tracked inside the operator [(link)](https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/blob/master/controllers/flinkcluster_observer.go#L284-L287) rather than the submitter script. And since the previous job tracking script had an issue like #294  with the flink operator's auto job recovery feature, this PR changed to track the job by ID unlike before. Could you explain more about your HA deployment?

Reply:
@abroskin The job is now being tracked inside the operator [(link)](https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/blob/master/controllers/flinkcluster_observer.go#L284-L287) rather than the submitter script. And since the previous job tracking script had an issue like #294 with the flink operator's auto job recovery feature, this PR changed to track the job by ID unlike before. And is the `spec.job.restartPolicy` set to `Never`? If so, the flink operator should not automatically restart the job. In addition, it would be nice if you could explain more about the HA configuration.